### PR TITLE
Add Semantic Similarity with BERT example

### PR DESCRIPTION
**Pull request by MohamadMerchant**

### CODE REVIEW

File: examples/nlp/semantic_similarity_with_bert.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+## Keras Custom Data Generator
+"""
+
+
+class DataGenerator(tf.keras.utils.Sequence):
+    """
+        Generates batch data for Keras.
+
+        Parameters:
+            sentence1: Text Input 1
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:141]: Argument meaning unclear

File: examples/nlp/semantic_similarity_with_bert.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            sentence2: Text Input 2
+            labels: Targets
+            batch_size: Size of the batch
+            shuffle: Whether to shuffle data or not
+            train: Use DataGenerator for train or test purpose
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:146]: This argument should be `include_targets`

File: examples/nlp/semantic_similarity_with_bert.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    valid_df.sentence2.astype("str"),
+    y_val,
+    batch_size=BATCH_SIZE,
+    shuffle=False,
+)
+
+"""
+## Train the Model
+"""
+h = model.fit_generator(
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:304]: Use full-spelled argument names

File: examples/nlp/semantic_similarity_with_bert.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    lambda x: 0 if x == "contradiction" else 1 if x == "entailment" else 2
+)
+y_test = tf.keras.utils.to_categorical(test_df.label, num_classes=3)
+
+"""
+## Keras Custom Data Generator
+"""
+
+
+class DataGenerator(tf.keras.utils.Sequence):
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:136]: Please use a more descriptive name than "DataGenerator"

File: examples/nlp/semantic_similarity_with_bert.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    proba = model.predict(test_data)[0]
+    idx = np.argmax(proba)
+    proba = f"{proba[idx]: .2f}%"
+    pred = LABELS[idx]
+    return pred, proba
+
+
+"""
+
+"""
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:346]: "Check results on some example sentence pairs"

File: examples/nlp/semantic_similarity_with_bert.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    """
+    input_ids = tf.keras.layers.Input(shape=(MAXLEN,), dtype=tf.int32, name="input_ids")
+    attention_masks = tf.keras.layers.Input(
+        shape=(MAXLEN,), dtype=tf.int32, name="att_mask"
+    )
+    token_type_ids = tf.keras.layers.Input(
+        shape=(MAXLEN,), dtype=tf.int32, name="tt_ids"
+    )
+    # Loading pretrained BERT model
+    bertModel = TFBertModel.from_pretrained(MODEL)
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:245]: Use lower case camel case for variable names

File: examples/nlp/semantic_similarity_with_bert.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+## Configurations
+"""
+
+MAXLEN = 128
+BATCH_SIZE = 32
+EPOCHS = 4
+LR = 3e-5
+np.random.seed(42)
+# we will use base-base-uncased pretrained model
+MODEL = "bert-base-uncased"
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:43]: "MODEL" is too generic, this is "pretrained_model_name"

File: examples/nlp/semantic_similarity_with_bert.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+
+"""
+# Recommended training on GPU Runtime
+try:
+    # build model with distributed strategy
+    strategy = tf.distribute.MirroredStrategy()
+    with strategy.scope():
+        model = build_model()
+    print(f"Strategy: {strategy}")
+except:
```
[the remark this answers is not in the document]
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: I thought it will fail while training on CPU but it doesn't. So I have removed try-except block as it doesn't seem necessary.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
+"""
+Title: Semantic Similarity with BERT
+Author: Mohamad Merchant
+Date created: 2020/08/15
+Last modified: 2020/08/15
+Description: Natural Language Inference by Fine tuning BERT model on SNLI Corpus.
+"""
+"""
+## **Introduction**
```
[the remark this answers is not in the document]
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Removed bold markers.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+valid_df = pd.read_csv("SNLI_Corpus/snli_1.0_dev.csv")
+test_df = pd.read_csv("SNLI_Corpus/snli_1.0_test.csv")
+
+# Shape of the data
+print(f"Total train samples : {train_df.shape[0]}")
+print(f"Total validation samples: {valid_df.shape[0]}")
+print(f"Total test samples: {valid_df.shape[0]}")
+
+"""
+  Dataset Info:
```
[the remark this answers is not in the document]
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Formatted.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+"""
+print(f"Sentence1: {train_df.loc[1, 'sentence1']}")
+print(f"Sentence2: {train_df.loc[1, 'sentence2']}")
+print(f"Similarity: {train_df.loc[1, 'similarity']}")
+
+"""
+## Preprocessing
+"""
+
+# we have some nan in our train data, we will simply drop them
```
[the remark this answers is not in the document]
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Done with all the comments.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+
+
+"""
+## Build The Model
+"""
+
+
+def build_model():
+    """
+    model inputs:
```
[the remark this answers is not in the document]
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Changed it to comment descriptions.

File: examples/nlp/semantic_similarity_with_bert.py (hunk 3 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Changed all argument/variable names as full-spelled

File: examples/nlp/semantic_similarity_with_bert.py (hunk 4 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Used BertSemanticDataGenerator.

File: examples/nlp/semantic_similarity_with_bert.py (hunk 5 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py:346]: Added line.

File: examples/nlp/semantic_similarity_with_bert.py (hunk 6 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Used lower_case for all variable names.

File: examples/nlp/semantic_similarity_with_bert.py (hunk 7 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: removed "MODEL"  from configurations and added description comment while loading.

**Comment by MohamadMerchant** [comment]
Thanks for the review @fchollet . I have made all the necessary changes.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+We will fine-tune BERT model by providing two sentences as inputs and it outputs the
+probability of the similarity between sentences.
+
+### References
+* [BERT](https://arxiv.org/pdf/1810.04805.pdf)
+* [SNLI](https://nlp.stanford.edu/projects/snli/)
+"""
+
+"""
+## Setup
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:24]: Please mention that `transformers` should be installed.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+"""
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+import transformers
+from transformers import BertTokenizer, TFBertModel
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:31]: You don't need this import line, you can just access these objects from `transformers`.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+
+"""
+## Configurations
+"""
+
+max_length = 128  # Maximum length of input sentence to the model.
+batch_size = 32
+epochs = 4
+learning_rate = 3e-5
+np.random.seed(42)
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:41]: It is unclear what this seed is meant to control. It will have no effect over Keras reproducibility. If it is specifically for the data shuffling, prefer passing a `seed` argument to your `Sequence` object and use it locally there (via a `RandomState` object). In general avoid global seeding since it has side effects.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+valid_df = (
+    valid_df[valid_df.similarity != "-"]
+    .sample(frac=1.0, random_state=42)
+    .reset_index(drop=True)
+)
+
+"""
+Distribution of our training and validation targets.
+"""
+print(train_df.similarity.value_counts())
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:107]: Use string formatting to print a sentence that makes it clear this is training (resp. validation) data

File: examples/nlp/semantic_similarity_with_bert.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+print(f"Sentence1: {train_df.loc[1, 'sentence1']}")
+print(f"Sentence2: {train_df.loc[1, 'sentence2']}")
+print(f"Similarity: {train_df.loc[1, 'similarity']}")
+
+"""
+## Preprocessing
+"""
+
+# We have some nan in our train data, we will simply drop them.
+print(train_df.isnull().sum())
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:89]: Use string formatting to print a sentence

File: examples/nlp/semantic_similarity_with_bert.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+    batch_size=batch_size,
+    shuffle=False,
+)
+
+"""
+## Train the Model
+"""
+history = model.fit_generator(
+    train_data,
+    steps_per_epoch=len(train_data) // batch_size,
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:301]: Since a `Sequence` has a `__len__`, this argument should not be necessary (same for validation_steps)

File: examples/nlp/semantic_similarity_with_bert.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+    y_test,
+    batch_size=batch_size,
+    shuffle=False,
+)
+model.evaluate_generator(test_data, steps=len(test_data) // batch_size, verbose=1)
+
+"""
+## Inference on custom sentences
+"""
+labels = ["CONTRADICTION", "ENTAILMENT", "NEUTRAL"]
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:322]: This is a kind of global constant, so you can move it higher up in the file

File: examples/nlp/semantic_similarity_with_bert.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+            "bert-base-uncased", do_lower_case=True
+        )
+        self.on_epoch_end()
+
+    def __len__(self):
+        # Denotes the number of batches per epoch.
+        return len(self.sentence1) // self.batch_size
+
+    def __getitem__(self, idx):
+        # Generates batch of data.
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:187]: You can be more precise: "Retrieves the batch of index `idx`."

File: examples/nlp/semantic_similarity_with_bert.py [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+    valid_df.sentence2.astype("str"),
+    y_val,
+    batch_size=batch_size,
+    shuffle=False,
+)
+
+"""
+## Train the Model
+"""
+history = model.fit_generator(train_data, validation_data=valid_data, epochs=epochs,)
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:306]: A big problem here is that you're not doing fine-tuning properly.

To do fine-tuning, you need 2 steps:

1. Freeze `bert_model` and fit the model. This is called "feature extraction": you just reuse the pretrained features without modifying them.
2. unfreeze `bert_model` and fit again (with a very small learning rate). This is called fine-tuning.

Step 2 may not be necessary -- for many applications, feature extraction is sufficient.

Also, please note that:

- You can pass a generator to `fit`. `fit_generator` is legacy.
- If you fit from a generator, you should generally use data multiprocessing to get better GPU utilization. See the docs for `fit()`.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+## Evaluate model on the test set
+"""
+test_data = BertSemanticDataGenerator(
+    test_df.sentence1.astype("str"),
+    test_df.sentence2.astype("str"),
+    y_test,
+    batch_size=batch_size,
+    shuffle=False,
+)
+model.evaluate_generator(test_data, verbose=1)
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:317]: Likewise, just `evaluate`

File: examples/nlp/semantic_similarity_with_bert.py [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+        # set to true if data generator is used for training/validation
+        if self.include_targets:
+            labels = np.array(self.labels[indexes], dtype="int32")
+            return [input_ids, masks, token_type_ids], labels
+        else:
+            return [input_ids, masks, token_type_ids]
+
+    def on_epoch_end(self):
+        # Shuffle indexes after each epoch if shuffle is set to True.
+        self.indexes = np.arange(len(self.sentence1))
```
[the remark this answers is not in the document]
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py:228]: Moving this line inside `if` will result in error message when `shuffle=False` as the validation and test generator objects cannot find self.indexes in this line` : 
`indexes = self.indexes[idx * self.batch_size : (idx + 1) * self.batch_size]`

File: examples/nlp/semantic_similarity_with_bert.py (hunk 20 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: Added

File: examples/nlp/semantic_similarity_with_bert.py (hunk 21 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: I have used only "feature extraction" and trained by freezing `bert_model` that gave me accuracy around 37%.
After some tuning, I experimented stacking a classification head below freezed bert_model :
1. stacking 2 dense layers with dropout gave accuracy around 45%.
2. stacking 1 Bi-LSTM layer with dropout gave accuracy around 69%.

`Accuracy of unfreeze bert_model (our current model) is around 87%.`

I have gone through this [paper](https://arxiv.org/pdf/1911.03090.pdf) yesterday as they have experimented with different datasets and models, I found this result related to our task.

```
Frozen up to  MNLI     MNLI-mm
0th           84.2      82.0
9th           82.0      82.4
12th          56.4      57.1
```
--------
I have changed fit_generator to fit and also used multiprocessing for better GPU utilization.

Comment (fchollet) [reply examples/nlp/semantic_similarity_with_bert.py]: It makes sense that you get better results by training the entire BERT model, but in this case it makes little sense to load a pretrained model (since you lose its representations almost immediately due to the presence of untrained layers in the model).

Either load the BERT model from scratch (no pretrained weights), or do actual fine-tuning (the two steps I described above). The first option is the simplest.

File: examples/nlp/semantic_similarity_with_bert.py [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+
+"""
+## Train the Model
+"""
+history = model.fit(
+    train_data,
+    validation_data=valid_data,
+    epochs=epochs,
+    use_multiprocessing=True,
+    workers=-1,
```
Comment (fchollet) [inline examples/nlp/semantic_similarity_with_bert.py:312]: Use 4

File: examples/nlp/semantic_similarity_with_bert.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/nlp/semantic_similarity_with_bert.py:228]: Since it is used in getitem, you should define self.indexes in __init__.

File: examples/nlp/semantic_similarity_with_bert.py (hunk 21 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py]: I got your point for the first time but I was little bit concerned about the accuracy. Now, the model is trained on both the steps  "feature extraction" and "fine-tuning" (unfreeze bert_model)  that gave quite good results.
I have also changed the data generator and used batch_encode_plus that has speed up the training.

File: examples/nlp/semantic_similarity_with_bert.py (hunk 24 above)
Comment (MohamadMerchant) [reply examples/nlp/semantic_similarity_with_bert.py:330]: Using 4 or 8 causes nondeterministic deadlocks and doesn't start the training.
(-1 uses 8 workers in colab.)

**Comment by MohamadMerchant** [comment]
Thank you for the great review, It was great contributing to Keras.
